File: app-common/lib/rest-api.cs
using Newtonsoft.Json;
using System;
using System.IO;
using System.Web;

namespace Rest {

  public class IInp<T, TData> {
    public string module;
    public string action;
    public T par;
    public string dataType;
  }

  public class IOut<T, TData> where T : class where TData : class {
    public string dataType;
    public T par;
    public string stringData;
    public byte[] binaryData;
  }

  public static class Types {
    public const string NO = "NO";
    public const string ERROR = "ERROR";
    public const string STRING = "STRING";
    public const string JSON = "JSON";
    public const string ARRAYBUFFER = "ARRAYBUFFER";
  }

  public class API : IHttpHandler {

    public void ProcessRequest(HttpContext context) {
      var headers = context.Request.Headers;
      var resp = context.Response;

      var module = headers["x-lm-module"];
      var type = headers["x-lm-action"];
      var dataType = headers["x-lm-data-type"];
      var par = headers["x-lm-par"];
      string stringData = null;
      BinaryReader binaryData = null;

      try {
        if (dataType != null)
          switch (dataType) {
            case Types.JSON:
            case Types.STRING:
              using (var reader = new StreamReader(context.Request.InputStream)) stringData = reader.ReadToEnd();
              break;
            default:
              binaryData = new BinaryReader(context.Request.InputStream);
              break;
          }

        IOut<Object, Object> res;

        switch (module) {
          case Recording.Consts.recording:
            var recRes = Recording.API(type, stringData, context.Server.MapPath("~/App_Data"));
            res = new IOut<Object, Object> { par = recRes.par, dataType = recRes.dataType, binaryData = recRes.binaryData, stringData = recRes.stringData };
            break;
          case "testModule":
            var testRes = TestRestAPI.API.testModule(type, par, dataType, null, binaryData);
            res = new IOut<Object, Object> { par = testRes.par, dataType = testRes.dataType, binaryData = testRes.binaryData, stringData = testRes.stringData };
            break;
          default:
            throw new Exception("Unknown module: " + module);
        }

        resp.ContentType = "application/octet-stream";
        resp.Headers["x-lm-type"] = res.dataType;
        if (res.par != null) resp.Headers["x-lm-par"] = HttpUtility.UrlEncode(JsonConvert.SerializeObject(res.par));
        if (res.stringData != null) resp.Write(res.stringData);
        if (res.binaryData != null) //BinaryReader reads this data type in little-endian format, https://stackoverflow.com/questions/9618608/is-net-binaryreader-always-little-endian-even-on-big-endian-systems
          using (var wr = new BinaryWriter(resp.OutputStream))
            foreach (var b in res.binaryData) wr.Write(b);
      } catch (Exception exp) {
        resp.Headers["x-lm-type"] = "ERROR";
        resp.Write(JsonConvert.SerializeObject(exp.Message));
      } finally {
        if (binaryData != null) binaryData.Close();
      }
    }

    public bool IsReusable { get { return false; } }

  }

}

File: app-common/lib/rest-api.ts
export const restAPI = async <T extends {} = {}, TData extends {} = {}>(data: RestAPI.IInp) => {

  return fetch(window.lmGlobal.platform.restAPIPlatform.serviceUrl, {
    method: 'post',
    headers: {
      'Content-Type': 'application/octet-stream',
      'x-lm-module': data.module,
      'x-lm-action': data.action,
      ...(data.dataType ? { 'x-lm-data-type': data.dataType } : null),
      ...(data.par ? { 'x-lm-par': encodeURIComponent(JSON.stringify(data.par)) } : null),
    },
    body: data.data && data.dataType === RestAPI.Types.JSON ? JSON.stringify(data.data) : data.data
  })
    .then(resp => {
      const out = {} as RestAPI.IOut<T, TData>
      out.dataType = resp.headers.get('x-lm-type') as RestAPI.Types 
      const par = resp.headers.get('x-lm-par')
      if (par) out.par = JSON.parse(decodeURIComponent(par))
      switch (out.dataType) {
        case RestAPI.Types.ERROR: throw new Error(resp.headers['x-lm-msg'])
        case RestAPI.Types.JSON: return resp.text().then(txt => {
          try { out.data = txt ? JSON.parse(txt) : null } catch (msg) { throw new Error(`ERROR: ${msg} (${txt})`) }
          return out
        })
        case RestAPI.Types.STRING: return resp.text().then(txt => { out.data = txt; return out })
        case RestAPI.Types.ARRAYBUFFER: return resp.arrayBuffer().then(buf => { out.data = buf; return out })
        case RestAPI.Types.NO: return Promise.resolve(out)
        default: throw 'not implemented'
      }
    })
}

function isLittleEndian(): boolean { //.NET binary writer is little endian
  var arrayBuffer = new ArrayBuffer(2)
  var uint8Array = new Uint8Array(arrayBuffer)
  var uint16array = new Uint16Array(arrayBuffer)
  uint8Array[0] = 0xAA // set first byte
  uint8Array[1] = 0xBB // set second byte
  return uint16array[0] === 0xBBAA 
}